boot/seal.go

```diff
@@ -83,3 +86,102 @@ func sealKeyToModeenv(key secboot.EncryptionKey, model *asserts.Model, modeenv *
 
 	return nil
 }
+
+type bootChain struct {
+	Model          string      `json:"model"`
+	BrandID        string      `json:"brand-id"`
+	Grade          string      `json:"grade"`
+	ModelSignKeyID string      `json:"model-sign-key-id"`
+	AssetChain     []bootAsset `json:"asset-chain"`
+	Kernel         string      `json:"kernel"`
+	// KernelRevision is the revision of the kernel snap. It is empty if
+	// kernel is unasserted, in which case always reseal.
+	KernelRevision string `json:"kernel-revision"`
+	KernelCmdline  string `json:"kernel-cmdline"`
+}
+
+type bootAsset struct {
+	Role   string   `json:"role"`
+	Name   string   `json:"name"`
+	Hashes []string `json:"hashes"`
+}
+
+// helper types
+type sortedHashesBootAsset bootAsset
+type sortedAssetsBootChain bootChain
+
+// copyForMarshalling copies the contents of boot asset and sorts the hash list.
+func (b *bootAsset) copyForMarshalling() *sortedHashesBootAsset {
+	newB := *b
+	newB.Hashes = make([]string, len(b.Hashes))
+	copy(newB.Hashes, b.Hashes)
+	sort.Strings(newB.Hashes)
+	return (*sortedHashesBootAsset)(&newB)
+}
+
+// MarshalJSON marshals the boot asset into a form that is deterministic and
+// suitable for equivalence tests.
+func (b *bootAsset) MarshalJSON() ([]byte, error) {
```

Comment:
I don't think we need this one, we won't marshal bootAsset alone outside of a bootChain. As is it will just make sort twice in the useful case?